velox/docs/develop/memory.rst

```diff
@@ -0,0 +1,890 @@
+       options.queryMemoryCapacity = queryMemoryGb << 30;
+       ...
+     }
+     memory::initializeMemoryManager(options);
+
+     if (systemConfig->asyncDataCacheEnabled()) {
+       ...
+       cache_ = std::make_shared<cache::AsyncDataCache>(
+          memoryManager()->allocator(), memoryBytes, std::move(ssd));
+     }
+     ...
+   }
+
+* L6: set the memory allocator capacity (system memory limit) from
+  Prestissimo system config
+* L7: set the memory allocator type from Prestissimo system config. If
+  *useMmapAllocator* is true, we use *MmapAllocator*, otherwise use
+  *MallocAllocator*. `Memory Allocator section <#memory-allocator>`_ describes these two
+  types of allocators
+* L9: set the memory arbitrator kind from the Prestissimo system config.
+  Currently, we only support the *“SHARED”* arbitrator kind (see `memory arbitrator section <#memory-arbitrator>`_).
+  *“NOOP”* arbitrator kind will be deprecated soon (`#8220 <https://github.com/facebookincubator/velox/issues/8220>`_)
+* L11: set the memory arbitrator capacity (query memory limit) from
+  Prestissimo system config
+* L14: creates the process-wide memory manager which creates memory
+  allocator and arbitrator inside based on MemoryManagerOptions initialized from previous steps
+* L15-19: creates the file cache if it is enabled in Prestissimo system
+  config
+
+Memory Pool
+-----------
+The memory pool provides memory allocation functions for query execution. It
+also tracks a query’s memory usage for per-query memory limit enforcement.
+As shown in the Query Memory Pool Hierarchy figure, a query creates a tree of
+memory pools that mirrors the query plan to have a fine-grained tracking of
+memory usage for figuring out which task(s) or operator(s) use most of the
+memory. At the root of the tree, *QueryCtx* creates a root query pool from the
+memory manager. Each query task creates a child task pool from the query pool.
+A query task executes a fragment of the query plan (e.g. an execution stage in
+a distributed query execution plan in Prestissimo). Each plan node in a task’s
+plan fragment creates a child node pool from the task pool
+(*Task::getOrAddNodePool*). Each plan node belongs to one or more task
+execution pipelines. Each pipeline might have multiple driver instances running
+in parallel. Each driver instance consists of a pipeline of query operators,
+and an operator is an instantiation of a query plan node in a driver. Hence
+each operator creates a child operator pool from the node pool
+(*Task::addOperatorPool*).
+
+.. image:: images/memory-pool.png
+    :width: 500
+    :align: center
+    :alt: Memory Pool
+
+Query allocates memory from the operator pool at the leaf of the tree and
+propagates the memory usage all the way up to the query pool at the root of the
+tree to check if the memory usage has exceeded the per-query memory limit or
+not. The memory allocation always happens at the leaf operator pool, the
```

Comment:
Might be helpful to discuss whether leaf memory pools are thread-safe or not.

Reply:
It is mentioned in memory usage tracking.